nodes/test/agent_rocketride/test_prune_wave_context.py

```
@@ -0,0 +1,264 @@
+        original_ids = [w['wave_num'] for w in waves]
+
+        driver._prune_wave_context(waves)
+
+        assert [w['wave_num'] for w in waves] == original_ids, (
+            'Master waves list was mutated by _prune_wave_context'
+        )
+
+    def test_master_list_never_mutated_by_budget_eviction(self) -> None:
+        """Character-budget eviction must only affect the returned copy."""
+        driver = _driver(context_window_waves=10, wave_context_budget_chars=150)
+        # 3 waves × 100 chars = 300 chars → budget eviction will fire
```

Ruff (0.15.12)

[warning] Comment contains ambiguous `×` (MULTIPLICATION SIGN). Did you mean `x` (LATIN SMALL LETTER X)?

(RUF003)